File: companion/bolumler/guvenirlik/ornek-01/cozum.R
d <- read.csv("veri.csv", na.strings=c("", "NA"))
stopifnot(identical(names(d),paste0("M",1:5)),!anyNA(d),all(as.matrix(d) %in% 1:5))
alpha<-function(x){k<-ncol(x);v<-var(rowSums(x));stopifnot(nrow(x)>1,k>1,v>0);k/(k-1)*(1-sum(sapply(x,var))/v)}
x<-d;x$M4<-6-x$M4
s<-c(n=nrow(x),alpha_ham=alpha(d),alpha=alpha(x),toplam_ortalama=mean(rowSums(x)))
for(col in names(x)){rest<-x[names(x)!=col];s<-c(s,setNames(c(cor(x[[col]],rowSums(rest)),alpha(rest)),paste0(col,c("_duzeltilmis_r","_silinirse_alpha"))))}
out<-data.frame(olcut=names(s),deger=as.numeric(s))
print(out,row.names=FALSE);write.csv(out,"sonuclar-r.csv",row.names=FALSE)
if("--check" %in% commandArgs(trailingOnly=TRUE)){
 ref<-read.csv("beklenen-sonuclar.csv")
 stopifnot(identical(out$olcut,ref$olcut),all(abs(out$deger-ref$deger)<=1e-10+1e-8*abs(ref$deger)))
 cat("KONTROL BASARILI\n")
}
sessionInfo()

File: companion/bolumler/guvenirlik/ornek-01/cozum.py
import argparse
from pathlib import Path
import numpy as np
import pandas as pd
from scipy import stats

parser = argparse.ArgumentParser()
parser.add_argument('--check', action='store_true')
args = parser.parse_args()
d = pd.read_csv('veri.csv')
assert list(d.columns)==['M1','M2','M3','M4','M5']
assert not d.isna().any().any() and d.isin([1,2,3,4,5]).all().all()
def alpha(x):
    k=x.shape[1]; v=x.sum(axis=1).var(ddof=1)
    if len(x)<2 or k<2 or v<=0: raise ValueError('Alfa için yeterli değişkenlik yok.')
    return k/(k-1)*(1-x.var(ddof=1).sum()/v)
x=d.copy(); x['M4']=6-x['M4']
s={'n':len(x),'alpha_ham':alpha(d),'alpha':alpha(x),'toplam_ortalama':x.sum(axis=1).mean()}
for col in x:
    rest=x.drop(columns=col)
    s[col+'_duzeltilmis_r']=x[col].corr(rest.sum(axis=1))
    s[col+'_silinirse_alpha']=alpha(rest)
out=pd.DataFrame(list(s.items()),columns=['olcut','deger'])
print(out.to_string(index=False))
out.to_csv('sonuclar-python.csv',index=False)
if args.check:
    ref=pd.read_csv('beklenen-sonuclar.csv')
    if list(ref.olcut)!=list(out.olcut): raise AssertionError('Sonuç etiketleri farklı.')
    np.testing.assert_allclose(out.deger,ref.deger,rtol=1e-8,atol=1e-10)
    print('KONTROL BAŞARILI')
